android/maplibreui/src/main/java/com/stadiamaps/ferrostar/maplibreui/extensions/VisualNavigationViewConfig.kt

```diff
@@ -0,0 +1,65 @@
+package com.stadiamaps.ferrostar.maplibreui.extensions
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.MutableState
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.platform.LocalLayoutDirection
+import androidx.compose.ui.unit.LayoutDirection
+import com.mapbox.mapboxsdk.geometry.LatLngBounds
+import com.maplibre.compose.camera.CameraState
+import com.maplibre.compose.camera.MapViewCamera
+import com.maplibre.compose.camera.models.CameraPadding
+import com.stadiamaps.ferrostar.composeui.config.CameraControlState
+import com.stadiamaps.ferrostar.composeui.config.VisualNavigationViewConfig
+import com.stadiamaps.ferrostar.core.NavigationUiState
+import com.stadiamaps.ferrostar.core.boundingBox
+import com.stadiamaps.ferrostar.maplibreui.NavigationViewMetrics
+
+@Composable
+fun VisualNavigationViewConfig.cameraControlState(
+    camera: MutableState<MapViewCamera>,
+    navigationCamera: MapViewCamera,
+    uiState: NavigationUiState,
+    navigationViewMetrics: NavigationViewMetrics
+): CameraControlState {
+  val cameraIsTrackingLocation = camera.value.state is CameraState.TrackingUserLocationWithBearing
+  val cameraControlState =
+      if (!cameraIsTrackingLocation) {
+        CameraControlState.ShowRecenter { camera.value = navigationCamera }
+      } else {
+        val bbox = uiState.routeGeometry?.boundingBox()
+        if (bbox != null) {
+          val scale = LocalDensity.current.density
+          val progressViewHeight = navigationViewMetrics.progressViewSize.height.value.toDouble()
+          val instructionsViewHeight =
+              navigationViewMetrics.instructionsViewSize.height.value.toDouble()
+          val layoutDirection = LocalLayoutDirection.current
+
+          // Bottom padding must take the recenter button into account
+          val bottomPadding = (progressViewHeight + this.buttonSize.height.value + 50) * scale
```

Comment:
Let it be known that I really don't like that we have to do this as a result of moving the button, further reducing the amount of space available for the map. I still think the UI would make more sense if we left the recenter button in the top end grid cell.

Reply:
Personally, I'd probably just ignore the button heights in the bounding box calculation. I think just filling the entire rectangle between the instructions and progress view with a reasonable padding probably succeeds in _most_ cases. The chances of a button hiding important route visual are already low & a user can easily adjust the map by pan if need be.